libyul/optimiser/VarNameCleaner.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/VarNameCleaner.h>
+#include <libyul/AsmData.h>
+#include <algorithm>
+#include <cctype>
+#include <climits>
+#include <iterator>
+#include <string>
+
+using namespace yul;
+using namespace std;
+
+// ensure the following transforms
+// - input: a, a_1, a_1_2
+//	 output: a, a_1, a_2
+// - input: a, a_1, a_1_2, a_2
+//   output: a, a_1, a_2, a_3
+// - input: a_15, a_17
+//   output: a, a_2
+//
+// - input: abi_decode_256
+//   output: abi_decode
+
+void VarNameCleaner::run(Block& _ast)
+{
+	VarNameCleaner{}(_ast);
+}
+
+void VarNameCleaner::operator()(VariableDeclaration& _varDecl)
+{
+	for (TypedName& typedName: _varDecl.variables)
+	{
+		if (auto newName = makeCleanName(typedName.name.str()))
+		{
+			printf("VariableDeclaration: %s -> %s\n", typedName.name.str().c_str(), newName->c_str());
+			typedName.name = YulString{*newName};
+		}
+		else
+			printf("VariableDeclaration: %s\n", typedName.name.str().c_str());
+	}
+
+	ASTModifier::operator()(_varDecl);
+}
+
+void VarNameCleaner::operator()(Identifier& _identifier)
+{
+	if (auto newName = getCleanName(_identifier.name.str()))
+		_identifier.name = YulString{*newName};
+}
+
+VarNameCleaner::VarNameCleaner()
+{
+	printf("VarNameCleaner.ctor\n");
+}
+
+// make identifier clean, remember new name for reuse, keep them unique and simple
+boost::optional<string> VarNameCleaner::makeCleanName(string const& _name)
+{
+	if (auto newName = stripSuffix(_name))
+	{
+		printf("makeCleanName.1: %s -> %s (# %zu)\n", _name.c_str(), newName->c_str(),
+				m_usedNames.count(*newName));
+		if (!m_usedNames.count(*newName))
+		{
+			printf("makeCleanName.1: %s -> %s\n", _name.c_str(), newName->c_str());
+			m_usedNames[_name] = *newName;
+			m_usedNames[*newName] = *newName;
+			return newName;
+		}
+
+		// create new name with suffix (by finding a free identifier)
+		for (uintmax_t i = 1; i < numeric_limits<decltype(i)>::max(); ++i)
```

Comment:
This seems overkill. size_t should cause memory exhaustion anyway and fit in a register.